container-hook: fix sys_exit_execve from thread

alban opened the pull request:
# container-hook: fix sys_exit_execve from thread

Similar to the builtin & image based trace-exec gadgets, the container-hook ebpf code uses sys_enter_execve and sys_exit_execve tracepoints to keep track of the command lines of executed processes.

Before this patch, this ebpf code had the same bug as in the trace-exec gadgets before (see https://github.com/inspektor-gadget/inspektor-gadget/pull/2454 and https://github.com/inspektor-gadget/inspektor-gadget/pull/2475).

sys_enter_execve adds an entry in the "exec_args" map and sys_exit_execve removes it. However, sys_enter_execve and sys_exit_execve can be called from different thread id in case of a multithreaded application. This caused the map to not get cleaned up properly.

This patch fixes the tracking between sys_enter_execve and sys_exit_execve.

Additionally, the kretprobe/fsnotify_remove_first_event needs to lookup the "exec_args" map by pid (and not tid) because IG does not use FAN_REPORT_TID (Linux >= 4.20). But the "exec_args" map is keyed by tid in order to be able to clean it up correctly. So kretprobe/fsnotify_remove_first_event also uses the pid_by_tgid ebpf map to do the conversion.

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/2789

## How to use

No changes in the UI.

## Testing done

Start ig with one of those commands:
```
go run -exec 'sudo -E' ./cmd/ig/... list-containers -w
sudo ./ig list-containers -w
```

In another terminal, check if the exec_args map gets leaked entries:
```
sudo sh -c 'while sleep 1 ; do bpftool map dump name exec_args|grep key|wc -l ; done'
```

Start the workload triggering the bug with one of those commands:
```
docker run --rm -ti albantest.azurecr.io/goexecve:t1 sh -c 'while true ; do goexecve /none/cat ; sleep 1 ; done'
kubectl run --image albantest.azurecr.io/goexecve:t1 --restart=Never goexecve2 -- sh -c 'while true ; do goexecve /none/cat ; sleep 1 ; done'
```

The workload is as follows:

<details>

```Dockerfile
FROM busybox
ADD goexecve /bin/
CMD while true ; do goexecve ; sleep 1 ; done
```

```Makefile
build:
	go build -o goexecve goexecve.go
```

```golang
package main

import (
	"os"
	"runtime"
	"syscall"
)

func main() {
	runtime.LockOSThread()
	go func() {
		runtime.LockOSThread()

		args := []string{"cat", "/dev/null"}
		env := os.Environ()

		binary := "/bin/cat"

		if len(os.Args) > 1 {
			binary = os.Args[1]
		}

		err := syscall.Exec(binary, args, env)
		if err == nil {
			panic("exec /none/cat should have failed")
		}

		os.Exit(0)
	}()
	select {}
}
```

</details>

mauriciovasquezbernal reviewed (commented):
I'm confused about what's the key of the exec_args map. 
Comment says `tid`, so I suppose it's `pid` from kernel point of view. But then, I see tgid is used to access it on L167 & L169. 

The trace exec gadget uses the pid:

https://github.com/inspektor-gadget/inspektor-gadget/blob/4f471903d5c9da31b1b8b046dc3b419dcab99c59/pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c#L108-L111

shouldn't it be doing the same?

alban commented on pkg/container-hook/bpf/execruntime.bpf.c:
```diff
@@ -31,10 +31,44 @@ struct {
 	__uint(type, BPF_MAP_TYPE_HASH);
 	__uint(max_entries, 128);
-	__type(key, u32); // tgid (fanotify will need to lookup by tgid)
+	__type(key, u32); // tid
```
| Kernel terminology   | Userspace terminology |
|----------------------|-----------------------|
| tgid (task group) | pid (process id)      |
| pid     | tid (thread id)       |

Should I use the kernel terminology in the code everywhere?

In the kernel terminology, this `exec_args` hash map was keyed by a tgid before the patch and I changed it to be keyed by a pid in order to fix the leak.

It makes things slightly more complicated in `kretprobe/fsnotify_remove_first_event` because the fanotify_event contains the tgid. So we have to use the `pid_by_tgid` map to find the pid, so we can do the lookup in `exec_args`.

mauriciovasquezbernal replied:
> Should I use the kernel terminology in the code everywhere?

I think it's a good idea to use kernel terminology on eBPF programs.

alban commented:
I rebased and fixed the following:
- tgid/pid mixups in L159, L167, L169, L224, L230
- changed the comment to use kernel terminology (`// pid (not tgid)`).
- fixed the two `__atomic_add_fetch`s in `sys_exit_execve`

It seems to work for me.

I tested that my programs `goexecve` generates the correct results and that the two maps `exec_args` and `pid_by_tgid` are empty after the completion of the execve.

mauriciovasquezbernal reviewed (approved):
LGTM. I can see how the map is not leaked with this change, and how it was leaked before.

Thanks for this fix!
